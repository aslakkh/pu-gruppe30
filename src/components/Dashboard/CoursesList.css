.CoursesList{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.ListElementHere{
    display: flex;
    flex-direction: row;
}
