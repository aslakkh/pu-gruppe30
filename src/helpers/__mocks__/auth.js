
export function auth (email, pw) {
    return true;
}

export function privilige(user, auth0) {
    return true;
}

export function logout () {
    return true;
}

export function login (email, pw) {
    return true;
}

export function saveUser (user) {
    return true;
}

export function saveCourse (course){
    return true;
}

//add course to courses/course
//for now sets all points to 7.5
export function addCourseToRoot(course){
    return true;
}

//check for existing course at root 
export function courseExistsAtRoot(course){
    return true;
}

export function getCourse(){
    return true;
}

//Function for getting all courses of a student (not fully implemented, see componentDidMount() in Courses.js)
export function getCourses(){
    return true;
}



export function getUserUid(){
    return true;
}

//function for removing course from user
export function removeCourse(course){
    return true;
}

export function removeCourseFromRoot(course){
    return true;
}

//function for setting a courses child 'active' to false
export function disableCourse(course){
    return true;
}

export function loadCourse(){
    return true;
}

export function planSession(course, date, goal){
    return true;
}


export function saveGoal(course, goal) {
    return true;
}
export function saveGoal2(view, course, seconds, date) {
    return true;
}


// saves goal and links seconds elapsed to a specific goal
//daily/monthly, weekly goal functions as active goal. Need a method that checks if goal expired
export function saveGoal3(goalType, course, goalInSeconds, secondsElapsed=null) {
    return true;
}

export function saveExpiredGoal(goalType, timeCreated, course, goalInSeconds, secondsSpent) {
    return true;
}

export function saveProgress(course, view, timeSpent) {
    return true;

}

/*
 export function gg(course, goalType, dateNow) {
 const userUid = firebase.auth().currentUser.uid;
 var goalRef = ref.child('users/'+userUid+"/courses/" + course +'/goals/'+ '/goalRegister/');
 var s;
 ratingRef.on("value", function (data) {
 s = data.val().weekly;
 });
 return s;
 }
 */

export function activateGoal(type, course, boolean) {
    return true;
}

export function test(course){
    return true;
}

export function getSec(course) {
    return true;
}

export function isGoalActive(course, view) {
    return true;
}