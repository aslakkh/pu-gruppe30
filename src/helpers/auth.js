import { ref, firebaseAuth } from '../config/constants'
import firebase from 'firebase';



export function auth (email, pw) {
    return firebaseAuth().createUserWithEmailAndPassword(email, pw)
        .then(saveUser)
        .catch((error) => console.log('Oops', error))
}

export function privilige(user, auth0) {
    console.log(user.uid);
    console.log(auth0);
    return ref.child(`users/${user.uid}/info`)
        .update({
            privilege: auth0
        })
}

export function logout () {
    return firebaseAuth().signOut()
}

export function login (email, pw) {
    return firebaseAuth().signInWithEmailAndPassword(email, pw)
}

export function saveUser (user) {
    return ref.child(`users/${user.uid}/info`)
        .set({
            email: user.email,
            uid: user.uid
        })
        .then(() => user)
}

export function saveCourse (course){
    const userUid = getUserUid();
    let usersRef = ref.child('users/'+userUid+"/courses/");
    let monthlyRef = ref.child('users/'+userUid+"/courses/" + course + "/goals/monthlyGoal/");
    let weeklyRef = ref.child('users/'+userUid+"/courses/" + course + "/goals/weeklyGoal/");
    let dailyRef = ref.child('users/'+userUid+"/courses/" + course + "/goals/dailyGoal/");

    usersRef.child(course).update({
        active: true
    });

    monthlyRef.update({
        active: true,
        timeSet: 0,
        value: 0,
        timeSpent: 0

    });

    weeklyRef.update({
        active: true,
        timeSet: 0,
        value: 0,
        timeSpent: 0
    });

    dailyRef.update({
        active: true,
        timeSet: 0,
        value: 0,
        timeSpent: 0
    });

    let goalRef = ref.child('users/'+userUid+"/courses/" + course + "/oldGoals/daily/1");
    goalRef.update({
        goal: 0,
        timeSet: 0,
        timeSpent: 0,
    });

    goalRef = ref.child('users/'+userUid+"/courses/" + course + "/oldGoals/weekly/1");
    goalRef.update({
        goal: 0,
        timeSet: 0,
        timeSpent: 0,
    });

    goalRef = ref.child('users/'+userUid+"/courses/" + course + "/oldGoals/monthly/1");
    goalRef.update({
        goal: 0,
        timeSet: 0,
        timeSpent: 0,
    });

}

//add course to courses/course
//for now sets all points to 7.5
export function addCourseToRoot(course){
    var courseRef = ref.child('courses/');
    courseRef.child(course).update({
        points: 7.5,
        weekly: 36000,
    })
}

//check for existing course at root 
export function courseExistsAtRoot(course){
    var courseRef = ref.child('courses/');
    var courseExists;
    courseRef.once('value', function(snapshot) {
        courseExists = snapshot.hasChild(course);
    });
    return courseExists;
}

export function getCourse(){
    const userUid = firebase.auth().currentUser.uid;
    this.firebase.database().ref(userUid).once();
    return firebase.database().ref(userUid).on("value", function(snapshot) {
        console.log(snapshot.val());
    });
}

//Function for getting all courses of a student (not fully implemented, see componentDidMount() in Courses.js)
export function getCourses(){
    const userUid = firebase.auth().currentUser.uid;
    //firebase.database().ref(userUid).once();
    return firebase.database().ref(userUid).child('courses').on("value", snap => {
        snap.val();
    });
}



export function getUserUid(){
    const userUid = firebase.auth().currentUser.uid;
    return userUid;
}

//function for removing course from user
export function removeCourse(course){
    const userUid = getUserUid();
    var userRef = ref.child('users/'+userUid+'/courses/');
    userRef.child(course).remove();
}

export function removeCourseFromRoot(course){
    var courseRef = ref.child('courses');
    courseRef.child(course).remove();
}

//function for setting a courses child 'active' to false
export function disableCourse(course){
    const userUid = getUserUid();
    var userRef = ref.child('users/'+userUid+'/courses/');
    userRef.child(course).update({
        active: false
    });
}

export function loadCourse(){
    console.log("const");
    var emner =[]
    const userUid = firebase.auth().currentUser.uid;
    ref.child('users/'+userUid+'/courses').once("value", function(snapshot){
        snapshot.forEach(function(data){
            console.log(data.val(),data.key)
            emner.push(data.key);

        })});
    return(emner);
}

export function planSession(course, date, goal){
    const userUid = getUserUid();
    var userRef = ref.child('users/'+userUid+'/courses/'+course+'/planned-sessions/');
    userRef.child(date).update({
        goal: goal,
    });
}


export function saveGoal(course, goal) {
    const userUid = firebase.auth().currentUser.uid;
    let courseRef = ref.child('users/'+userUid+'/courses/'+course);
    courseRef.update({
        goal:goal
    });
}
export function saveGoal2(view, course, seconds, date) {
    const userUid = firebase.auth().currentUser.uid;
    let courseRef = ref.child('users/'+userUid+'/courses/'+course + '/goals/' + view);
    courseRef.update({
        value:seconds,
        timeSet:date
    });
}


// saves goal and links seconds elapsed to a specific goal
//daily/monthly, weekly goal functions as active goal. Need a method that checks if goal expired
export function saveGoal3(goalType, course, goalInSeconds, secondsElapsed=null) {
    const userUid = firebase.auth().currentUser.uid;
    const timeRef = ref.child('users/'+userUid+'/courses/'+ course +'/goals/'+ '/goalRegister/' +  Date.now());
    timeRef.set({
        goalType: goalType,
        goalInSeconds: goalInSeconds,
        secondsElapsed: secondsElapsed
    });
}

export function saveExpiredGoal(goalType, timeCreated, course, goalInSeconds, secondsSpent) {
    const userUid = firebase.auth().currentUser.uid;
    console.log("halloiluken")
    console.log('users/'+userUid+'/courses/'+ course + '/oldGoals/' + goalType + "/" + timeCreated)
    const goalRef = ref.child('users/'+userUid+'/courses/'+ course + '/oldGoals/' + goalType + "/" + timeCreated);
    goalRef.set({
        goal: goalInSeconds,
        timeSet: timeCreated,
        timeSpent: secondsSpent
    });
}

export function saveProgress(course, view, timeSpent) {
    const userUid = firebase.auth().currentUser.uid;
    const goalRef = ref.child('users/'+userUid+'/courses/'+ course +'/goals/'+ view);
    goalRef.update({
        timeSpent: timeSpent
    });

}

/*
 export function gg(course, goalType, dateNow) {
 const userUid = firebase.auth().currentUser.uid;
 var goalRef = ref.child('users/'+userUid+"/courses/" + course +'/goals/'+ '/goalRegister/');
 var s;
 ratingRef.on("value", function (data) {
 s = data.val().weekly;
 });
 return s;
 }
 */

export function activateGoal(type, course, boolean) {
    const userUid = firebase.auth().currentUser.uid;
    let courseRef = ref.child('users/'+userUid+'/courses/'+course + '/goals/' + type);
    courseRef.update({
        active: boolean
    });
}

export function test(course){
    //const userUid = firebase.auth().currentUser.uid;
    //firebase.database().ref(userUid).once();
    var s;
    let coursesRef = ref.child("/courses/" + course + "/weekly");
    coursesRef.on("value", function(snapshot){
        s = snapshot.val();
    });
    return s;
}

export function getSec(course) {
    var ratingRef = firebase.database().ref("courses/" + course);
    var s;
    ratingRef.on("value", function (data) {
        s = data.val().weekly;
    });
    return s;
}

export function isGoalActive(course, view) {
    const userUid = firebase.auth().currentUser.uid;
    let out = 0;
    let coursesRef = ref.child('users/'+userUid+'/courses/' + course + "/goals/" + view + "/active/");
    coursesRef.on("value", function(snapshot){
        out = snapshot.val();
    });
    return out;
}


export function removeOldGoal(course, type, key) {
    const userUid = firebase.auth().currentUser.uid;
    let oldGoalRef = ref.child('users/'+userUid+'/courses/'+course + '/oldGoals/' + type + "/");
    console.log(oldGoalRef);
    oldGoalRef.child(key).remove();
}
