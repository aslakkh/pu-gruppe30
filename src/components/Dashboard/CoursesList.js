import React, { Component } from 'react'
import {ListGroup, ListGroupItem, Button} from 'react-bootstrap'
import {disableCourse} from '../../helpers/auth'
import {styles} from './CoursesListStyles.js'


/*      Component for displaying a users courses   

Returns courses as ListGroup
Renders upon change in DB

*/ 


export default class CoursesList extends Component {

	constructor(props) {
		super(props);
		this.state={
			courses: this.props.courses
		}
	}


	componentWillReceiveProps(nextProps){
		this.setState({
				courses: nextProps.courses,
			});
		
}



	handleClick(key){
		disableCourse(key); //sets "active" to false in database
	}



    //Component renders a header, followed by the keys of json object retrieved from firebase mapped to list
	render(){
		if(this.state.courses === undefined || this.state.courses === null){
			return(
				<h5>You have no active courses. Add a course above.</h5>
			);
		}
		else{
			return(
				<ListGroup>
						{Object.keys(this.state.courses).map((key) => {
							return (
								<ListGroupItem key={key} style={styles.coursesList}> 
									<div style={styles.listElement}>{key}</div>
									<Button bsStyle="danger" onClick={() => this.handleClick(key)}>Delete</Button>
								</ListGroupItem>
							);
								
						})} 
				</ListGroup>
		);
		}
	}

}


