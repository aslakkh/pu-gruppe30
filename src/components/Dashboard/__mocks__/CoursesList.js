import React, { Component } from 'react'
import {ListGroup, ListGroupItem, Button} from 'react-bootstrap'
import {disableCourse, courseExistsAtRoot} from '../../helpers/auth'
import {CourseWarning} from '../various/CourseWarning'
import {styles} from './CoursesListStyles.js'


/*      Component for displaying a users courses   

Returns courses as ListGroup
Renders upon change in DB

TODO: 
- Ask for confirmation on Delete
- Submit with enter
- Add courses in chronological order?
- ()
Could probably be split into a Course-component for each courses
*/ 

//http://stackoverflow.com/questions/40494787/get-user-object-properties-from-firebase-in-react

export default class CoursesList extends Component {

	constructor(props) {
		super(props);
		this.state={
			
			courses: this.props.courses
		}
	}

	componentDidMount(){
		//should be implemented in auth.js
		/*var self = this;
		var userUid = getUserUid(); //gets userid of user currently logged in

        //following code gets users courses from firebase and updates the components state
        //Should maybe be implemented in a function in auth.js
		self.firebaseRef = firebase.database().ref().child("users/"+userUid+"/courses");
		self.firebaseRef.orderByChild("active").equalTo(true).on('value', snap => {
			console.log(snap.val());
			self.setState({
				
				courses: snap.val() //sets courses to whatever is at users/userUid/courses in database
			});
		});*/
		
	}

	componentWillReceiveProps(nextProps){
		this.setState({
			courses: nextProps.courses
		});

}

	componentWillUnmount(){

		//this.firebaseRef.off('value');
	}

	//handles click
	handleClick(key){
		return true;
	}



    //Component renders a header, followed by the keys of json object retrieved from firebase mapped to list
	render(){
		
		if(this.state.courses === undefined || this.state.courses === null){
			return(
				<h5>You have no active courses. Add a course below.</h5>
			);
		}
		else{
			return(
				<ListGroup>
						{Object.keys(this.state.courses).map((key) => {
							
							return <ListGroupItem key={key} /*className="CoursesList"*/ style={styles.coursesList}> 
								<div style={styles.listElement}>{key}</div>
								<Button bsStyle="danger" onClick={() => this.handleClick(key)}>Delete</Button>
								</ListGroupItem>
			
							
						})} 
				</ListGroup>
		);
		}
	}

}


