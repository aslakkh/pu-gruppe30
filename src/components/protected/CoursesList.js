import React, { Component } from 'react'
import {ListGroup, ListGroupItem, Button} from 'react-bootstrap'
import { getUserUid, removeCourse} from '../../helpers/auth'
import firebase from 'firebase'
import './CoursesList.css'


/*      Component for displaying a users courses   

Returns courses as ListGroup
Renders upon change in DB

TODO: 
FIX ERROR REMOVING LAST ELEMENT IN LIST / LOADING A USER WITH NO COURSES
FIX WARNING MESSAGE ON LOAD AFTER VIEWING COURSES
Could probably be split into a Course-component for each courses
*/ 

//http://stackoverflow.com/questions/40494787/get-user-object-properties-from-firebase-in-react



export default class CoursesList extends Component {

	constructor(props) {
		super(props);
		this.state={
			courses: []
		}
	}

	componentDidMount(){
		//should be implemented in auth.js
		var self = this;
		var userUid = getUserUid(); //gets userid of user currently logged in

        //following code gets users courses from firebase and updates the components state
        //Should maybe be implemented in a function in auth.js
		const coursesRef = firebase.database().ref().child("users/"+userUid+"/courses");
		coursesRef.on('value', snap => {
			self.setState({
				courses: snap.val() //sets courses to whatever is at users/userUid/courses in database
			});
		});
	}

	//handles click
	handleClick(key){
		removeCourse(key);
	}



    //Component renders a header, followed by the keys of json object retrieved from firebase mapped to list
	render(){
		

		
		return(
			<div>
					
					<ListGroup >
							{Object.keys(this.state.courses).map((key) => {
								return <ListGroupItem key={key} className="CoursesList"> 
										{key} 
										<Button bsStyle="danger" onClick={() => this.handleClick(key)}>Delete</Button>
										</ListGroupItem>;
							})} 
					</ListGroup>
			</div>
		);
		
	}

}


