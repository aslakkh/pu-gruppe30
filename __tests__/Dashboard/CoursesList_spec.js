import React from 'react';
import {shallow} from 'enzyme';
import TestUtils from "react-addons-test-utils";
import auth from '../../src/helpers/auth.js';
import CoursesList from '../../src/components/Dashboard/CoursesList.js';


jest.mock('../../src/helpers/auth.js');

const courses = {
    IT5555: {
        active: true,
        goal: 0,
        goals: {},
        oldGoals: {},
        time: 0,
    },
    TDT0000: {
        active: true,
        goal: 0,
        goals: {},
        oldGoals: {},
        time: 0,
    }
}

const course = {
    IT5555: {
        active: true,
        goal: 0,
        goals: {},
        oldGoals: {},
        time: 0,
    },
}

describe("CoursesList", function(){
    var renderer;
    renderer = TestUtils.createRenderer();

    renderer.render(
        <CoursesList />
    )
    var component = renderer.getRenderOutput();

    it('renders without props', () => {
        expect(component).toMatchSnapshot();
    })

    renderer.render(
        <CoursesList courses={courses}/>
    )
    var componentWithProps = renderer.getRenderOutput();

    it('renders with props', () => {
        expect(componentWithProps).toMatchSnapshot();
    })

    it('handles click', () => {

        const myMock = jest.fn();
        const test = shallow(
            <CoursesList courses={course}/>
        );
        test.find('Button').simulate('click');
    })

})
