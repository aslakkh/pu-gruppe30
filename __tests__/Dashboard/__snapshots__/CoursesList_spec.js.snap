// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesList renders with props 1`] = `
<ListGroup
  bsClass="list-group"
>
  <ListGroupItem
    bsClass="list-group-item"
    listItem={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      IT5555
    </div>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="danger"
      disabled={false}
      onClick={[Function]}
    >
      Delete
    </Button>
  </ListGroupItem>
  <ListGroupItem
    bsClass="list-group-item"
    listItem={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      TDT0000
    </div>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="danger"
      disabled={false}
      onClick={[Function]}
    >
      Delete
    </Button>
  </ListGroupItem>
</ListGroup>
`;

exports[`CoursesList renders without props 1`] = `
<h5>
  You have no active courses. Add a course above.
</h5>
`;
